### DISCUSSION TITLE
i am stuck with a load of errors please help

### DISCUSSION THREAD

**Discussion by Akash47007** [micropython/micropython#15636] [Using MicroPython]
I don't know what is wrong but I get these errors recursively

UnicodeError
OSError: [Errno 104] ECONNRESET
OSError: [Errno 103] ECONNABORTED

`import utime
import network
import usocket
from machine import Pin, SPI
import ili9341
import xpt2046

# Initialize the ILI9341 display
class main():
    def __init__ (self):
        SSID = "xxx"
        PASSWORD = "xxx"

        # MySQL server credentials
        self.DB_HOST = "192.168.0.xxx"  # Replace with your MySQL server's IP address
        self.DB_USER = "xxx"
        self.DB_PASSWORD = "xxx"
        self.DB_NAME = "study_assistant_db"

        # Connect to Wi-Fi
        self.station = network.WLAN(network.STA_IF)
        self.station.active(True)
        self.station.connect(SSID, PASSWORD)

        while not self.station.isconnected():
            utime.sleep(1)
        print("Wi-Fi connected:", self.station.ifconfig())
        
        spi1 = SPI(0,
                  baudrate=40000000,
                  polarity=1,
                  phase=1,
                  bits=8,
                  firstbit=SPI.MSB,
                  sck=Pin(18),
                  mosi=Pin(19),
                  miso=Pin(16))
        self.display = ili9341.Display(spi1, dc=Pin(15), cs=Pin(17), rst=Pin(14))
        spi2 = SPI(1, baudrate=1000000, sck=Pin(10), mosi=Pin(11), miso=Pin(12))

        # Initialize touch screen
        self.int = Pin(9)
        self.touch = xpt2046.Touch(spi2, cs=Pin(13), int_pin=self.int,int_handler=self.touchscreen_press)
        
        self.main_menu()

    def connect_to_db(self):
        addr = usocket.getaddrinfo(self.DB_HOST, 3306)[0][-1]
        self.s = usocket.socket()
        self.s.connect(addr)
        self.s.send(f"CONNECT {self.DB_USER}@{self.DB_HOST} IDENTIFIED BY '{self.DB_PASSWORD}'\n".encode())
        self.s.send(f"USE {self.DB_NAME}\n".encode())
        return self.s

    def execute_query(self,socket, query):
        socket.send(query.encode())
        return socket.recv(256)  # Reduce buffer size for query results

    def clear_screen(self):
        self.display.clear()

    def display_text(self,text, x, y, color):
        self.display.draw_text8x8(x, y,text,ili9341.color565(*color))

    def draw_button(self,x, y, w, h, color, text):
        self.display.draw_rectangle(x, y, w, h, ili9341.color565(*color))
        self.display_text(text, x + 10, y + (h // 2) - 4, (255, 255, 255))

    def dropdown_menu(self,options, x, y, color):
        height = 20
        for i, option in enumerate(options):
            self.display.draw_rectangle(x, y + i * height, 160, height, ili9341.color565(*color))
            self.display_text(option, x + 10, y + i * height + 4, (255, 255, 255))

    def main_menu(self):
        self.clear_screen()
        self.display_text("Study Assistant", 60, 20, (255, 255, 0))
        self.draw_button(40, 80, 160, 40, (0, 100, 200), "Start Session")
        self.draw_button(40, 140, 160, 40, (0, 200, 100), "Stop Session")
        self.draw_button(40, 200, 160, 40, (200, 0, 0), "View Sessions")

    def touchscreen_press(self,x,y):
            if 40 <= x <= 200 and 80 <= y <= 120:
                self.start_study_session()
            elif 40 <= x <= 200 and 140 <= y <= 180:
                self.stop_study_session()
            elif 40 <= x <= 200 and 200 <= y <= 240:
                self.view_study_sessions()

    def start_study_session(self):
        self.clear_screen()
    
        users = ["User1", "User2", "User3"]
        subjects = ["Math", "Science", "History"]
    
        self.display_text("Select User:", 60, 40, (255, 255, 255))
        self.dropdown_menu(users, 40, 60, (50, 50, 50))
    
        self.display_text("Select Subject:", 60, 120, (255, 255, 255))
        self.dropdown_menu(subjects, 40, 140, (50, 50, 50))
    
    # Start time
        start_time = utime.localtime()
    
    # Display ongoing session
        self.display_text("Session Started", 60, 180, (0, 255, 0))
    
    # Keep updating the duration until stop is called
        while True:
            duration = utime.time() - utime.mktime(start_time)
            minutes = duration // 60
            seconds = duration % 60
            self.display_text(f"Time: {minutes:02}:{seconds:02}", 60, 200, (255, 255, 255))
            utime.sleep(1)
            if self.int.value() == 1:  # A basic stop mechanism
                break
        
        
        self.db_socket = self.connect_to_db()
        query = f"INSERT INTO sessions (user_id, subject, start_time, actual_duration) VALUES ('{selected_user}', '{selected_subject}', '{start_time[3]:02}:{start_time[4]:02}:{start_time[5]:02}', {duration});"
        self.result = execute_query(db_socket, query)
    
        if "ERROR" not in result.decode():
            self.display_text("Session Saved", 60, 220, (0, 255, 0))
        else:
            self.display_text("Error Saving", 60, 220, (255, 0, 0))
    
        utime.sleep(2)
        self.main_menu()

    def stop_study_session(self):
        self.display_text("Session Stopped", 60, 200, (255, 0, 0))
        utime.sleep(2)
        self.main_menu()

    def view_study_sessions(self):
        self.clear_screen()
        self.display_text("Fetching Sessions...", 60, 120, (255, 255, 0))
    
        self.db_socket = self.connect_to_db()
        query = "SELECT user_id, subject, actual_duration FROM study_log;"
        result = self.execute_query(self.db_socket, query)
    
        if result:
            y_offset = 10
            print(result)
            dcr = result.decode()
            sessions = dcr.splitlines()
            for session in sessions:
                self.display_text(session, 10, y_offset, (255, 255, 255))
                y_offset += 20
                if y_offset > 240:
                    break  # Display as many sessions as fit on the screen
        else:
            self.display_text("Error Loading", 60, 200, (255, 0, 0))
    
        utime.sleep(2)
        self.main_menu()

# Run the main menu
main()

`

**Comment by peterhinch** [comment]
Please format your code according to [the guidelines](https://github.com/orgs/micropython/discussions/9111). It would also help anyone reading your code if you provided URLs for your external modules (ili9341, xpt2046).

Have you tested your SQL query at the REPL? I haven't used SQL for years so I may well be wrong - but the syntax looks odd to me.

I'd also avoid using `int` as the name of a bound variable as it's a Python keyword.

**Comment by Akash47007** [comment]
If someone has adequate knowledge, please help

**Reply by peterhinch** [reply]
People might be more motivated to help if you responded to suggestions already made.

As I said above, the SQL queries could be tested at the REPL. For example
```py
query = f"INSERT INTO sessions (user_id, subject, start_time, actual_duration) VALUES ('{selected_user}', '{selected_subject}', '{start_time[3]:02}:{start_time[4]:02}:{start_time[5]:02}', {duration});"
```
looks wrong to me. Did you intend to use f-strings?

**Comment by Akash47007** [comment]
`
import utime 
import network
import usocket
from machine import Pin, SPI
import ili9341
import xpt2046 
`
`
class main () :

    def __init__ (self):

         # Initialize the ILI9341 display
        SSID = "xxxx"
        PASSWORD = "xxx"

        # MySQL server credentials
        self.DB_HOST = "xxxx"  # Replace with your MySQL server's IP address
        self.DB_USER = "client"
        self.DB_PASSWORD = "xxxx"
        self.DB_NAME = "study_assistant_db"

        # Connect to Wi-Fi
        self.station = network.WLAN(network.STA_IF)
        self.station.active(True)
        self.station.connect(SSID, PASSWORD)

        while not self.station.isconnected():
            utime.sleep(1)
        print("Wi-Fi connected:", self.station.ifconfig())
        
        spi1 = SPI(0,
                  baudrate=40000000,
                  polarity=1,
                  phase=1,
                  bits=8,
                  firstbit=SPI.MSB,
                  sck=Pin(18),
                  mosi=Pin(19),
                  miso=Pin(16))
        self.display = ili9341.Display(spi1, dc=Pin(15), cs=Pin(17), rst=Pin(14))
        spi2 = SPI(1, baudrate=1000000, sck=Pin(10), mosi=Pin(11), miso=Pin(12))

        # Initialize touch screen
        self.int = Pin(9)
        self.touch = xpt2046.Touch(spi2, cs=Pin(13), int_pin=self.int,int_handler=self.touchscreen_press)
        
        self.main_menu()

    def connect_to_db(self):
        addr = usocket.getaddrinfo(self.DB_HOST, 3306)[0][-1]
        self.s = usocket.socket()
        self.s.connect(addr)
        self.s.send(f"CONNECT {self.DB_USER}@{self.DB_HOST} IDENTIFIED BY '{self.DB_PASSWORD}'\n".encode())
        self.s.send(f"USE {self.DB_NAME}\n".encode())
        return self.s

    def execute_query(self,socket, query):
        socket.send(query.encode())
        return socket.recv(256)  # Reduce buffer size for query results

    def clear_screen(self):
        self.display.clear()

    def display_text(self,text, x, y, color):
        self.display.draw_text8x8(x, y,text,ili9341.color565(*color))

    def draw_button(self,x, y, w, h, color, text):
        self.display.draw_rectangle(x, y, w, h, ili9341.color565(*color))
        self.display_text(text, x + 10, y + (h // 2) - 4, (255, 255, 255))

    def dropdown_menu(self,options, x, y, color):
        height = 20
        for i, option in enumerate(options):
            self.display.draw_rectangle(x, y + i * height, 160, height, ili9341.color565(*color))
            self.display_text(option, x + 10, y + i * height + 4, (255, 255, 255))

    def main_menu(self):
        self.clear_screen()
        self.display_text("Study Assistant", 60, 20, (255, 255, 0))
        self.draw_button(40, 80, 160, 40, (0, 100, 200), "Start Session")
        self.draw_button(40, 140, 160, 40, (0, 200, 100), "Stop Session")
        self.draw_button(40, 200, 160, 40, (200, 0, 0), "View Sessions")

    def touchscreen_press(self,x,y):
            if 40 <= x <= 200 and 80 <= y <= 120:
                self.start_study_session()
            elif 40 <= x <= 200 and 140 <= y <= 180:
                self.stop_study_session()
            elif 40 <= x <= 200 and 200 <= y <= 240:
                self.view_study_sessions()

    def start_study_session(self):
        self.clear_screen()
    
        users = ["User1", "User2", "User3"]
        subjects = ["Math", "Science", "History"]
    
        self.display_text("Select User:", 60, 40, (255, 255, 255))
        self.dropdown_menu(users, 40, 60, (50, 50, 50))
    
        self.display_text("Select Subject:", 60, 120, (255, 255, 255))
        self.dropdown_menu(subjects, 40, 140, (50, 50, 50))
    
    # Start time
        start_time = utime.localtime()
    
    # Display ongoing session
        self.display_text("Session Started", 60, 180, (0, 255, 0))
    
    # Keep updating the duration until stop is called
        while True:
            duration = utime.time() - utime.mktime(start_time)
            minutes = duration // 60
            seconds = duration % 60
            self.display_text(f"Time: {minutes:02}:{seconds:02}", 60, 200, (255, 255, 255))
            utime.sleep(1)
            if self.int.value() == 1:  # A basic stop mechanism
                break
        
        
        self.db_socket = self.connect_to_db()
        query = f"INSERT INTO sessions (user_id, subject, start_time, actual_duration) VALUES ('{selected_user}', '{selected_subject}', '{start_time[3]:02}:{start_time[4]:02}:{start_time[5]:02}', {duration});"
        self.result = execute_query(db_socket, query)
    
        if "ERROR" not in result.decode():
            self.display_text("Session Saved", 60, 220, (0, 255, 0))
        else:
            self.display_text("Error Saving", 60, 220, (255, 0, 0))
    
        utime.sleep(2)
        self.main_menu()

    def stop_study_session(self):
        self.display_text("Session Stopped", 60, 200, (255, 0, 0))
        utime.sleep(2)
        self.main_menu()

    def view_study_sessions(self):
        self.clear_screen()
        self.display_text("Fetching Sessions...", 60, 120, (255, 255, 0))
    
        self.db_socket = self.connect_to_db()
        query = "SELECT user_id, subject, actual_duration FROM study_log;"
        result = self.execute_query(self.db_socket, query)
    
        if result:
            y_offset = 10
            print(result)
            dcr = result.decode()
            sessions = dcr.splitlines()
            for session in sessions:
                self.display_text(session, 10, y_offset, (255, 255, 255))
                y_offset += 20
                if y_offset > 240:
                    break  # Display as many sessions as fit on the screen
        else:
            self.display_text("Error Loading", 60, 200, (255, 0, 0))
    
        utime.sleep(2)
        self.main_menu()

main()

`

**Comment by sophiedegran** [comment]
Hi Akash. why you don't use try/ except on your defs to have more lisibility? it's hard to reproduce and to answer on your code.. 
and for me it will be usefull to use async as you have many 'process'

**Comment by shariltumin** [comment]
Under ```while True``` block in "# Keep updating the duration until stop is called" you have:

```
    self.db_socket = self.connect_to_db()
    query = f"INSERT INTO sessions (user_id, subject, start_time, actual_duration) VALUES ('{selected_user}', '{selected_subject}', '{start_time[3]:02}:{start_time[4]:02}:{start_time[5]:02}', {duration});"
    self.result = execute_query(db_socket, query)
```

The ```self.result = execute_query(db_socket, query)``` 
should be ```self.result = execute_query(self.db_socket, query)``` 
since "db_socket" is not defined.

I do not understand why you have to do ```self.db_socket = self.connect_to_db()``` every time you want to do ```self.execute_query()```, unless you have closed the DB server socket by calling self.db_socket.close().

You can just do the "connect_to_db()" once and store the socket in self.db_socket after you have done the DB authentication.

Anyway, "UnicodeError" is probably related to str.encode() and result.decode(). While the "ECONNRESET" (connection reset by DB server) and ECONNABORTED (your connection closed due to error).

You can fix your script if you can find the places in your code where these errors occurred.

My advice is to start with a simple script that does only
1. Connect to the DB server port, i.e. get a client socket
2. Send authentication i.e. CONNECT ...
3. Connect to a database i.e. USE ...
4. Send a query
5. Read the response

Repeat 4 and 5 several times.
